Support custom HTTP/Socket options.

davisp opened the pull request:
This is a port of COUCHDB-1008 to chttpd. It provides a "server_options"
and "socket_options" variable in the [chttpd] section of the config that
will affect the Mochiweb server and individual connection sockets.

[1] https://github.com/apache/couchdb/commit/ac02c7bdca
[2] https://issues.apache.org/jira/browse/COUCHDB-1008

davisp committed:
Support custom HTTP/Socket options.

This is a port of COUCHDB-1008 to chttpd. It provides a "server_options"
and "socket_options" variable in the [chttpd] section of the config that
will affect the Mochiweb server and individual connection sockets.

[1] https://github.com/apache/couchdb/commit/ac02c7bdca
[2] https://issues.apache.org/jira/browse/COUCHDB-1008

kocolosk commented on src/chttpd.erl:
```diff
@@ -123,6 +126,8 @@ config_change("chttpd", "bind_address") ->
     ?MODULE:stop();
 config_change("chttpd", "backlog") ->
+    ?MODULE:stop();
+config_change("chttpd", "socket_options") ->
```
Shouldn't this be `server_options`?  The `socket_options` are queried on every request.

davisp replied:
That was an easter egg to test if anyone actually reviewed these things and was in no way an unintentional brain fart.

I swears it.

kocolosk commented:
LGTM, let's see what the Bobs think about the preference ordering of the various ways to set the server options in the morning.

bdionne commented:
LGTM

davisp commented:
@bdionne

No preference on the preference for option importance or deprecation of backlog?

Not really. Unless there are strong reasons otherwise, it should be consistent with couchdb, which I think it is. We may want to also add some brief comments in the .ini that explain what these options do

rnewson commented:
Let's remove chttpd backlog, including the two lines above where we listen for a change to it.

As for the preference order, could you explain the consequences (pre-caffeinated bob here)? It feels like the .ini file should take precedence but you did the opposite, explainz?

rnewson commented:
on the preference order, I think it's right as it is now. you were referring to the settings at the top of start_link, which were clipped in the diff.

kocolosk commented:
Ok, so regarding the preference order thing, I don't see anyone needing to change `loop` or `name` via server config.  Changing the IP address is absolutely something people might want to do.  Currently that's only possible via `[chttpd] bind_address = 127.0.0.1` and not via `[chttpd] server_options = [{ip, "127.0.0.1"}]`.  It's absolutely inconsistent but I guess it's consistent with Apache CouchDB.  `backlog` was never an Apache CouchDB option so we have more leeway to remove it.

kocolosk committed:
Remove support for [chttpd] backlog

Follow Apache CouchDB by using the server_options setting instead:

[chttpd]
server_options = [{backlog, 128}]

kocolosk commented:
Reviewing the CouchDB implementation it seems that setting `{ip, blah}` in the `server_options` would override the `bind_address` setting.  I think the solution proposed here is more sensible.

I'd say let's get this merged in, though we should probably squash @davisp's fixup into the original commit first.  I'd like to keep the removal of the support for `[chttpd] backlog` separate.

kocolosk commented:
Did what I suggested.  Merging.
